### PR TITLE
Case Insensitive Url

### LINKED ISSUE #556: Drop case sensitivity in URLs
lbry.io/reefermadness and lbry.io/ReeferMadness should both show the same page

Additionally, this ticket should not be considered complete without `rel=canonical` meta tag so that both count the same for SEO purposes.

### PR DESCRIPTION
**Pull request by maximest-pierre**
Fixes #556 #504

### CODE REVIEW

File: controller/Controller.class.php [hunk 1]
```diff
@@ -148,6 +152,10 @@ protected static function getRouterWithRoutes(): \Routing\RouteCollector
                 Response::enableHttpCache();
                 return ['page/' . $slug, []];
             } else {
+                $lower_slug = strtolower($slug);
+                if ($slug !== $lower_slug) {
+                    return static::execute("GET", $lower_slug);
```
Comment (kauffj) [inline controller/Controller.class.php]: `Route::GET`

File: controller/Controller.class.php [hunk 2]
```diff
@@ -148,6 +152,10 @@ protected static function getRouterWithRoutes(): \Routing\RouteCollector
                 Response::enableHttpCache();
                 return ['page/' . $slug, []];
             } else {
+                $lower_slug = strtolower($slug);
```
Comment (kauffj) [inline controller/Controller.class.php]: DRY violation

File: controller/Controller.class.php [hunk 3]
```diff
@@ -54,6 +54,10 @@ public static function execute($method, $uri)
             $dispatcher = new Routing\Dispatcher($router->getData());
             return $dispatcher->dispatch($method, $uri);
         } catch (\Routing\HttpRouteNotFoundException $e) {
+            $lower_uri = strtolower($uri);
```
Comment (kauffj) [inline controller/Controller.class.php]: capitalization doesn't match rest of file

File: controller/Controller.class.php [hunk 4]
```diff
@@ -148,13 +148,22 @@ protected static function getRouterWithRoutes(): \Routing\RouteCollector
...
             } else {
-                return NavActions::execute404();
+                return static::checkLowerUri($slug);
             }
         });
 
         return $router;
     }
 
+    protected static function checkLowerUri($uri)
```
Comment (kauffj) [inline controller/Controller.class.php]: This function has unclear side effects. One would not reasonably expect `checkLowerUri` to re-route to a 404.

There is likely a better name than this, but if you can't think of anything better, name it `doLowerUriRedirectOr404`.

**Comment by kauffj** [comment]
@NetOperatorWibby is correct that this still has problems.

Originally I was thinking we'd redirect to the lowercased version, but since we're just re-routing internally, this design results in multiple URLs serving the same content. So either you need to set a flag when this happens and set the meta tag appropriately, or you need to redirect (301) rather than just re-execute routing.

With regards to tokens, this is "not safe" in the sense that theoretically this could cause a spurious match, but it is safe in the sense that multiplying the tokens tried by two should have no real-world consequence given a sufficiently large number/token space (which we should have in all cases).

File: controller/Controller.class.php [hunk 5]
```diff
@@ -148,13 +148,24 @@ protected static function getRouterWithRoutes(): \Routing\RouteCollector
...
         });
 
         return $router;
     }
 
+    protected static function doLowerUriRedirectOr404($uri, $router = null)
+    {
+        $lowerUri = strtolower($uri);
+        if ($uri !== $lowerUri) {
+            if (View::exists('page/' . $lowerUri) || isset($router) && $router.hasroute($lowerUri)) {
```
Comment (kauffj) [inline controller/Controller.class.php]: Did you run this? I don't think `$router.hasroute` is valid PHP

Comment (maximest-pierre) [reply controller/Controller.class.php]: yes,
In route_collector.class.php which is the base class for $router:
```
public function hasRoute(string $name): bool
    {
        return isset($this->reverse[$name]);
    }
```

Comment (kauffj) [reply controller/Controller.class.php]: So you're confidant this code works?

Comment (maximest-pierre) [reply controller/Controller.class.php]: yes, I tested 404, dynamic and static routes.

**Comment by tzarebczan** [comment]
Putting this on hold until either @nikooo777 or @NetOperatorWibby can get to it, unless someone else wants to take over.

**Comment by nikooo777** [comment]
@NetOperatorWibby is this issue still valid with the new .io ?

**Comment by kauffj** [comment]
Yes, it's just a new design not new (backend) code.

**Pull request closed without merging**